mergify_engine/tasks/engine/__init__.py

```diff
@@ -161,24 +161,73 @@ def run(event_type, data):
         return
 
     with github.get_client(owner, repo, installation) as client:
-        _run(client, event_type, data)
+        pull = get_github_pull_from_event(client, event_type, data)
 
+        if not pull:  # pragma: no cover
+            LOG.info(
+                "No pull request found in the event %s, ignoring",
+                event_type,
+                gh_owner=client.owner,
+                gh_repo=client.repo,
+            )
+            return
+
+        # Get a fresh pull request view
+        # Everything after this must handle batching per pull request
+        sources = [{"event_type": event_type, "data": data}]
+        ctxt = context.Context(client, pull)
+
+        issue_comment_sources = []
+
+        for source in sources:
+            if source["event_type"] == "issue_comment":
+                issue_comment_sources.append(source)
+            elif not is_source_obsolete(ctxt.pull, source):
+                ctxt.sources.append(source)
+
+        commands_runner.spawn_pending_commands_tasks(ctxt)
+
+        if issue_comment_sources:
+            for source in issue_comment_sources:
+                commands_runner.run_command(
+                    ctxt,
+                    source["data"]["comment"]["body"],
+                    source["data"]["comment"]["user"],
+                )
 
-def _run(client, event_type, data):
-    raw_pull = get_github_pull_from_event(client, event_type, data)
+        if ctxt.sources:
+            run_actions(ctxt)
 
-    if not raw_pull:  # pragma: no cover
-        LOG.info(
-            "No pull request found in the event %s, ignoring",
-            event_type,
-            gh_owner=client.owner,
-            gh_repo=client.repo,
+
+def is_source_obsolete(pull, source):
+    event_type = source["event_type"]
+    data = source["data"]
+
+    if (
+        event_type == "status" and pull["head"]["sha"] != data["sha"]
+    ):  # pragma: no cover
+        context.get_logger(pull).debug(
+            "No need to proceed queue (got status of an old commit)", event_type,
         )
-        return
+        return True
+    elif (
+        event_type in ["status", "check_suite", "check_run"] and pull["merged"]
+    ):  # pragma: no cover
+        context.get_logger(pull).debug(
+            "No need to proceed queue (got status of a merged pull request)"
```

Comment:
That does not mean there is not a rule for this though. It sounds like a little premature optimization, no?
(I realize this is already in the code though so we can discuss that later)

Reply:
I hesitated to removed all of that, it's represent only 1500 events per days. And about this one in particular, maybe people quand status/check rule after pull have been merged.

Reply:
I decided to remove them.